hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/artifacts/upload/AbstractFileTransferHandler.java

```diff
@@ -180,59 +193,89 @@ protected void assertStateConsistency(final FileUploadId fileUploadId, final Str
         }
     }
 
-    protected OutputStream createOutputStreamForTempFile() throws IOException {
-
-        if (isUploadInterrupted()) {
-            return ByteStreams.nullOutputStream();
+    protected static void tryToCloseIOStream(final OutputStream outputStream) {
+        try {
+            outputStream.close();
+        } catch (final IOException e1) {
+            LOG.error("Closing output stream caused an exception {}", e1);
         }
-
-        final File tempFile = File.createTempFile("spUiArtifactUpload", null);
-
-        // we return the outputstream so we cannot close it here
-        @SuppressWarnings("squid:S2095")
-        final OutputStream out = new FileOutputStream(tempFile);
-
-        tempFilePath = tempFile.getAbsolutePath();
-
-        return out;
     }
 
-    protected String getTempFilePath() {
-        return tempFilePath;
+    protected static void tryToCloseIOStream(final InputStream inputStream) {
+        try {
+            inputStream.close();
+        } catch (final IOException e1) {
+            LOG.error("Closing input stream caused an exception {}", e1);
+        }
     }
 
-    // Exception squid:S3655 - Optional access is checked in
-    // checkIfArtifactDetailsDispalyed subroutine
-    @SuppressWarnings("squid:S3655")
-    protected void transferArtifactToRepository(final FileUploadId fileUploadId, final long fileSize,
-            final String mimeType, final String tempFilePath) {
-
-        final File newFile = new File(tempFilePath);
-
-        final String filename = fileUploadId.getFilename();
-        LOG.info("Transfering tempfile {} - {} to repository", filename, tempFilePath);
-        try (FileInputStream fis = new FileInputStream(newFile)) {
+    private final class TransferArtifactToRepositoryRunnable implements Runnable {
+        private final InputStream inputStream;
+        private final FileUploadId fileUploadId;
+        private final String mimeType;
+        private final UI vaadinUi;
+
+        public TransferArtifactToRepositoryRunnable(final InputStream inputStream, final FileUploadId fileUploadId,
+                final String mimeType, final UI vaadinUi) {
+            this.inputStream = inputStream;
+            this.fileUploadId = fileUploadId;
+            this.mimeType = mimeType;
+            this.vaadinUi = vaadinUi;
+        }
 
-            artifactManagement.create(fis, fileUploadId.getSoftwareModuleId(), filename, null, null, true, mimeType,
-                    fileSize);
+        @Override
+        public void run() {
+            try {
+                UI.setCurrent(vaadinUi);
+                streamToRepository();
+            } catch (final RuntimeException e) {
+                publishUploadFailedEvent(fileUploadId, i18n.getMessage("message.upload.failed"), e);
+                tryToCloseIOStream(inputStream);
```

Comment:
Why wouldn't you close the stream within a finally block to also cover the success case? I guess a try-with-resources statement cannot be used here...